hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OmSnapshotManager.java

```diff
@@ -298,27 +376,144 @@ private void verifySnapshotInfoForSnapDiff(final SnapshotInfo fromSnapshot,
+    } catch (RocksDBException exception) {
+      // TODO: [SNAPSHOT] Fail gracefully.
+      throw new RuntimeException(exception);
+    }
+  }
 
-    String dbPath = Paths.get(dbDirPath.toString(), OM_SNAPSHOT_DIFF_DB_NAME)
-        .toFile()
-        .getAbsolutePath();
+  /**
+   * Return the column family from column family list if it was existing
+   * column family, otherwise create new column family.
+   * This is for backward and forward compatibility.
+   * For backward compatibility, when column family doesn't exist. it will
+   * create new one and return that.
+   * For forward compatibility, it will return the existing one.
+   */
+  private ColumnFamilyHandle getOrCreateColumnFamily(
+      final String columnFamilyName,
+      final List<ColumnFamilyDescriptor> familyDescriptors,
+      final List<ColumnFamilyHandle> familyHandles) {
+
+    for (int i = 0; i < familyDescriptors.size(); i++) {
+      String cfName = StringUtils.bytes2String(familyDescriptors.get(i)
+          .getName());
+      if (columnFamilyName.equals(cfName)) {
+        return familyHandles.get(i);
+      }
+    }
 
     try {
-      return ManagedRocksDB.open(managedOptions, dbPath);
+      ColumnFamilyDescriptor columnFamilyDescriptor =
+          new ColumnFamilyDescriptor(StringUtils.string2Bytes(columnFamilyName),
+              columnFamilyOptions);
+      ColumnFamilyHandle columnFamily = snapshotDiffDb.get()
+          .createColumnFamily(columnFamilyDescriptor);
+
+      // Add column family and descriptor so that they can be closed if needed.
+      familyHandles.add(columnFamily);
+      familyDescriptors.add(columnFamilyDescriptor);
+      return columnFamily;
     } catch (RocksDBException exception) {
-      // TODO: Fail gracefully.
+      // TODO: [SNAPSHOT] Fail gracefully.
       throw new RuntimeException(exception);
     }
   }
 
+  private void closeRocksDbObjects(
+      final ManagedDBOptions managedDBOptions,
+      final ManagedColumnFamilyOptions managedColumnFamilyOptions,
+      final List<ColumnFamilyDescriptor> columnFamilyDescriptors,
+      final List<ColumnFamilyHandle> columnFamilyHandles,
+      final ManagedRocksDB managedRocksDB) {
+
+    if (managedDBOptions != null) {
+      managedDBOptions.close();
+    }
+    if (managedColumnFamilyOptions != null) {
+      closeColumnFamilyOptions(managedColumnFamilyOptions);
+    }
+    if (columnFamilyDescriptors != null) {
+      columnFamilyDescriptors.forEach(columnFamilyDescriptor ->
+          closeColumnFamilyOptions((ManagedColumnFamilyOptions)
+              columnFamilyDescriptor.getOptions()));
+    }
+    if (columnFamilyHandles != null) {
+      columnFamilyHandles.forEach(ColumnFamilyHandle::close);
+    }
+    if (managedRocksDB != null) {
+      managedRocksDB.close();
+    }
+  }
+
+  private void closeColumnFamilyOptions(
+      final ManagedColumnFamilyOptions managedColumnFamilyOptions) {
+    if (managedColumnFamilyOptions.isReused()) {
+      return;
```

Comment:
I guess this is a different problem but there is nowhere that sets:

```
options.setReused(false);
```

only `true`: https://github.com/apache/ozone/blob/4748fb97269526ecf67138fcd406ad380932e3ca/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/db/DatanodeDBProfile.java#L129-L133

Reply:
I was following [RocksDatabase](https://github.com/apache/ozone/blob/4748fb97269526ecf67138fcd406ad380932e3ca/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RocksDatabase.java#L166). It is true that we are not setting `options.setReused(false);` but I think closing is fine.

As you said it is different problem. Also in this case we are not even setting it to `true`.

Reply:
We could throw exception here instead, if Snapshot DB CFs won't ever be opened with `setReused(true)`

Reply:
Added the check.